libs/filer/workspace_fuse_client.go

```diff
@@ -0,0 +1,218 @@
+	oi1 workspace.ObjectInfo
+	oi2 workspace.ObjectInfo
+
+	commonName string
+}
+
+func (e DuplicatePathError) Error() string {
+	return fmt.Sprintf("cannot read files from the workspace file system by emulating FUSE. Duplicate paths encountered. Both %s at %s and %s at %s resolve to the same name %s", e.oi1.ObjectType, e.oi1.Path, e.oi2.ObjectType, e.oi2.Path, e.commonName)
+}
+
+// This is a filer for the workspace file system that allows you to pretend the
+// workspace file system is a traditional file system. It allows you to list, read, write,
+// delete, and stat notebooks (and files in general) in the workspace, using their paths
+// with the extension included.
+//
+// The ReadDir method returns a DuplicatePathError if this traditional file system view is
+// not possible. For example, a python notebook called foo and a python file called foo.py
+// would resolve to the same path foo.py in a tradition file system.
+//
+// Users of this filer should be careful when using the Write and Mkdir methods.
+// The underlying import API we use to upload notebooks and files returns opaque internal
+// errors for namespace clashes (e.g. a file and a notebook or a directory and a notebook).
+// Thus users of these methods should be careful to avoid such clashes.
+func NewWorkspaceFuseClient(w *databricks.WorkspaceClient, root string) (Filer, error) {
+	wc, err := NewWorkspaceFilesClient(w, root)
+	if err != nil {
+		return nil, err
+	}
+
+	return &workspaceFuseClient{
+		workspaceFilesClient: wc,
+	}, nil
+}
+
+func (w *workspaceFuseClient) ReadDir(ctx context.Context, name string) ([]fs.DirEntry, error) {
+	entries, err := w.workspaceFilesClient.ReadDir(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the entries by name to ensure that the order and any resultant
+	// errors are deterministic.
+	slices.SortFunc(entries, func(a, b fs.DirEntry) int {
+		return strings.Compare(a.Name(), b.Name())
+	})
+
+	seenPaths := make(map[string]workspace.ObjectInfo)
+	for i, entry := range entries {
+		info, err := entry.Info()
+		if err != nil {
+			return nil, err
+		}
+		sysInfo := info.Sys().(workspace.ObjectInfo)
+
+		// Skip if the object is not a notebook
+		if sysInfo.ObjectType != workspace.ObjectTypeNotebook {
+			if _, ok := seenPaths[entries[i].Name()]; ok {
+				return nil, DuplicatePathError{
+					oi1:        seenPaths[entries[i].Name()],
+					oi2:        sysInfo,
+					commonName: path.Join(name, entries[i].Name()),
+				}
+			}
+			seenPaths[entry.Name()] = sysInfo
+			continue
+		}
+
+		// Add extension to local file path if the file is a notebook
+		newEntry := entry.(wsfsDirEntry)
+		newEntry.wsfsFileInfo.oi.Path = newEntry.wsfsFileInfo.oi.Path + notebook.GetExtensionByLanguage(&sysInfo)
+		entries[i] = newEntry
+
+		if _, ok := seenPaths[newEntry.Name()]; ok {
+			return nil, DuplicatePathError{
+				oi1:        seenPaths[newEntry.Name()],
+				oi2:        sysInfo,
+				commonName: path.Join(name, newEntry.Name()),
+			}
+		}
```

Comment:
If this is possible in the workspace filesystem, we should remove the "fallback" entry, and just return.

There will be an issue committing this situation to Git as well. I don't think this needs to be the place to surface that error. We can just skip the fallback in this case and emit a log message.

Reply:
What do you mean by remove the fallback entry? Like omit the notebook entirely from the `ReadDir` output?

Reply:
Yeah -- we wouldn't be able to read it anyway.